### ISSUE TITLE
Elide lifetime in trait

### ISSUE THREAD

**Issue by amab8901** [rust-lang/rust-clippy#10394]
### Summary

Lifetime could be elided in the trait, but Clippy didn't notice that. Please add correction to clippy so that it warns about it and suggests to elide it

### Lint Name

needless_lifetimes

### Reproducer

I tried this code (copy+paste into main() function):

```rust
    #[derive(Debug)]
    pub enum Key<'a> {
        Ref(&'a [u8]),
        Val8([u8; 1]),
        Val16([u8; 2]),
        Val32([u8; 4]),
        Val64([u8; 8]),
        Val128([u8; 16]),
    }
    
    impl<'a> AsRef<[u8]> for Key<'a> {
    fn as_ref(&self) -> &[u8] {
        match self {
            Key::Ref(r) => r,
            Key::Val8(v) => v,
            Key::Val16(v) => v,
            Key::Val32(v) => v,
            Key::Val64(v) => v,
            Key::Val128(v) => v,
        }
    }
}
    
    pub trait Prefixer<'a> {
        /// returns 0 or more namespaces that should be length-prefixed and concatenated for range searches
        fn prefix(&self) -> Vec<Key>;
    
        // fn joined_prefix(&self) -> Vec<u8> {
        //     let prefixes = self.prefix();
        //     namespaces_with_key(&prefixes.iter().map(Key::as_ref).collect::<Vec<_>>(), &[])
        // }
    }
    impl<'a> Prefixer<'a> for Vec<u8> {
        fn prefix(&self) -> Vec<Key> {
            vec![Key::Ref(self.as_ref())]
        }
    }
```

I expected to see this happen:
Clippy should warn about this paragraph:
```
impl<'a> Prefixer<'a> for Vec<u8> {
    fn prefix(&self) -> Vec<Key> {
        vec![Key::Ref(self.as_ref())]
    }
}
```
Clippy should suggest to replace it with the following:
```
impl Prefixer<'_> for Vec<u8> {
    fn prefix(&self) -> Vec<Key> {
        vec![Key::Ref(self.as_ref())]
    }
}
```
Note that the changes are in the first line of the quoted paragraph, and particularly the lifetime parameters being simplified from two instances of `<'a>` into just one instance of `<'_>` 

Instead, this happened:
clippy had nothing to say about this code

### Version

```text
rustc -Vv
rustc 1.68.0-nightly (0442fbabe 2023-01-10)
binary: rustc
commit-hash: 0442fbabe24ec43636a80ad1f40a0ad92a2e38df
commit-date: 2023-01-10
host: x86_64-unknown-linux-gnu
release: 1.68.0-nightly
LLVM version: 15.0.6
```

**Comment by HKalbasi** [comment]
It also doesn't warn for inherent impls:
```Rust
struct S<'a>(&'a i32);
impl<'a> S<'a> {}
```

**Comment by samueltardieu** [comment]
> Clippy should suggest to replace it with the following:
> 
> ```
> impl Prefixer<'_> for Vec<u8> {
>     fn prefix(&self) -> Vec<Key> {
>         vec![Key::Ref(self.as_ref())]
>     }
> }
> ```
> 
> Note that the changes are in the first line of the quoted paragraph, and particularly the lifetime parameters being simplified from two instances of `<'a>` into just one instance of `<'_>`

It does now:

```
warning: the following explicit lifetimes could be elided: 'a
  --> /tmp/t.rs:35:6
   |
35 | impl<'a> Prefixer<'a> for Vec<u8> {
   |      ^^           ^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
help: elide the lifetimes
   |
35 - impl<'a> Prefixer<'a> for Vec<u8> {
35 + impl Prefixer<'_> for Vec<u8> {
```

**Comment by samueltardieu** [comment]
> It also doesn't warn for inherent impls:
> 
> struct S<'a>(&'a i32);
> impl<'a> S<'a> {}

It now does as well:

```
warning: the following explicit lifetimes could be elided: 'a
 --> /tmp/u.rs:4:6
  |
4 | impl<'a> S<'a> {}
  |      ^^    ^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
  = note: `#[warn(clippy::needless_lifetimes)]` on by default
help: elide the lifetimes
  |
4 - impl<'a> S<'a> {}
4 + impl S<'_> {}
```

**Comment by samueltardieu** [comment]
Closing, since the issue has been fixed.
